xplat: sync to latest and fix bugs

jianchun opened the pull request:
Sync to latest `chakracore-master` and `chakracore`.

Interesting changes in the last 2 commits:
- Expose chakracore version.
- Use experimental String APIs to fix String encoding conversion bugs.

whitlockjc committed:
timers: fix processing of nested timers

Whenever a timer is scheduled within another timer, there are a few
known issues that we are fixing:

* Whenever the timer being scheduled has the same timeout value as the
outer timer, the newly created timer can fire on the same tick of the
event loop instead of during the next tick of the event loop
* Whenever a timer is added in another timer's callback, its underlying
timer handle will be started with a timeout that is actually incorrect

This commit consists of
https://github.com/nodejs/node-v0.x-archive/pull/17203 and
https://github.com/nodejs/node-v0.x-archive/pull/25763.

Fixes: https://github.com/nodejs/node-v0.x-archive/issues/9333
Fixes: https://github.com/nodejs/node-v0.x-archive/issues/15447
Fixes: https://github.com/nodejs/node-v0.x-archive/issues/25607
Fixes: https://github.com/nodejs/node/issues/5426
PR-URL: https://github.com/nodejs/node/pull/3063

mcollina committed:
win,msi: Added Italian translation

As titled. Tested by @piccoloaiutante.

PR-URL: https://github.com/nodejs/node/pull/4647
Refs: https://github.com/wixtoolset/wix3/pull/366
Reviewed-By: Rod Vagg <rod@vagg.org>
Reviewed-By: Frederic Hemberger <mail@frederic-hemberger.de>

bnoordhuis committed:
deps: cherry-pick d721121 from v8 upstream

Original commit message:

    Quit creating array literal boilerplates from Crankshaft.

    It's such a corner case.

    BUG=

    Review URL: https://codereview.chromium.org/1865013002

    Cr-Commit-Position: refs/heads/master@{#35346}

Fixes: https://github.com/nodejs/node/issues/7454
PR-URL: https://github.com/nodejs/node/pull/7632
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>

lance committed:
doc: Warn against `uncaughtException` dependency.

State in the documentation that `uncaughtException` is not a reliable
way to restart a crashed application, and clarify that an application
may crash in ways that do not trigger this event.

Use a documented synchronous function in example code.

Fixes: https://github.com/nodejs/node/issues/6223

Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
PR-URL: https://github.com/nodejs/node/pull/6378

mcollina committed:
dgram: generalized send queue to handle close

If the udp socket is not ready and we are accumulating
messages to send, it needs to delay closing the socket when
all messages are flushed.

Fixes: https://github.com/nodejs/node/issues/7061
PR-URL: https://github.com/nodejs/node/pull/7066
Reviewed-By: Anna Henningsen <anna@addaleax.net>

Commit:
inspector: Unify event queues

Current implementation tracks connected/disconnected status separately
which potentially introduces race condition.
This change introduces notion of session IDs and also posts
connect/disconnect events into the same queue as the messages. This way
Node knows what session given response belongs to and can discard
messages if the frontend for that session had disconnected.

This also fixes an issue when frontend was unable to attach to V8
instance that was running infinite loop.

PR-URL: https://github.com/nodejs/node/pull/7271
Reviewed-By: bnoordhuis - Ben Noordhuis <info@bnoordhuis.nl>

santigimeno committed:
doc,dgram: fix addMembership documentation

Adding membership using `IP_ADD_MEMBERSHIP` with interface address set
to `INADDR_ANY` for `IPv4` or as an index of `0` for `IPv6` leads to
using only one interface selected by the operating system.

Fixes: https://github.com/nodejs/node/issues/1692
PR-URL: https://github.com/nodejs/node/pull/7244
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

ryanmurakami committed:
doc: grammar fixes to event loop guide

PR-URL: https://github.com/nodejs/node/pull/7479
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

thefourtheye committed:
net: export isIPv4, isIPv6 directly from cares

The function objects encapsulating `isIPv4` and `isIPv6` are not
necessary. They can be directly exposed from `cares`.

PR-URL: https://github.com/nodejs/node/pull/7481
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>

targos committed:
build: add v8 requirement to test-v8* in Makefile

The test targets expect that V8 is built in deps/v8/out

Ref: https://github.com/nodejs/node/issues/7477
PR-URL: https://github.com/nodejs/node/pull/7482
Reviewed-By: Myles Borins <myles.borins@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

thefourtheye committed:
tools: remove unnecessary imports and assignments

PR-URL: https://github.com/nodejs/node/pull/7483
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>

targos committed:
build: use BUILDTYPE when building V8 in Makefile

Without this it would always compile Release and Debug builds.

Ref: https://github.com/nodejs/node/issues/7477
PR-URL: https://github.com/nodejs/node/pull/7482
Reviewed-By: Myles Borins <myles.borins@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

indutny committed:
buffer: fix unintended unsigned overflow

`offset` is user supplied variable and may be bigger than
`ts_obj_length`. There is no need to subtract them and pass along, so
just throw when the subtraction result would overflow.

PR-URL: https://github.com/nodejs/node/pull/7494
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>

indutny committed:
crypto: fix undefined behavior in ParseExtension

Many extensions are unknown to the `ClientHelloParser::ParseExtension`,
do not cast user-supplied `uint16_t` to `enum`.

PR-URL: https://github.com/nodejs/node/pull/7494
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>

claudiorodriguez committed:
fs: rename event to eventType in fs.watch listener

The name 'event' for the argument of the listener in
fs.watch was confusing considering FSWatcher also had
events. This changes the name of the argument to
eventType.

Fixes: https://github.com/nodejs/node/issues/7504
PR-URL: https://github.com/nodejs/node/pull/7506
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>

diosney committed:
repl: add support for custom completions

Allow user code to override the default `complete()` function from
`readline.Interface`. See:
https://nodejs.org/api/readline.html#readline_use_of_the_completer_function

Ref: https://github.com/nodejs/node-v0.x-archive/pull/8484

PR-URL: https://github.com/nodejs/node/pull/7527
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Lance Ball <lball@redhat.com>

qheaden committed:
doc: dns.resolve fix callback argument description

The dns.resolve documentation stated that an array of IP
addresses would be returned in the callback. This is true
for everything other than the SOA record which returns an object.
This fixes that documentation.

Fixes: https://github.com/nodejs/node/issues/6506
PR-URL: https://github.com/nodejs/node/pull/7532
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Roman Reiss <me@silverwind.io>

diosney committed:
doc: added information on how to run the linter.

Added clarification about the linter execution.

PR-URL: https://github.com/nodejs/node/pull/7534
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

bnoordhuis committed:
build: add --enable-d8 configure option

Add an option to the configure script for building d8.  Useful for
testing V8 standalone.

PR-URL: https://github.com/nodejs/node/pull/7538
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rod Vagg <rod@vagg.org>

vsemozhetbyt committed:
doc: fix typo in the CHANGELOG_V6

compliment -> complement

PR-URL: https://github.com/nodejs/node/pull/7568
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

ChALkeR committed:
build: respect --shared-* flags for inspector deps

Don't build openssl/http_parser/libuv for v8_inspector if corresponding
--shared-* flags were passed to the ./configure script.

Fixes: https://github.com/nodejs/node/issues/7478
Fixes: https://github.com/nodejs/node/issues/7583
PR-URL: https://github.com/nodejs/node/pull/7569
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Ali Ijaz Sheikh <ofrobots@google.com>

Trott committed:
test: fix flaky test-fs-read-buffer-tostring-fail

The test is memory intensive and times out occasionally on Raspberry Pi
devices in CI. Successful test runs take about 90 seconds, but the
devices time out after 120 seconds. That's not a lot of headroom. So
let's skip the test on devices that have only modest amounts of memory.

Fixes: https://github.com/nodejs/node/issues/7042
PR-URL: https://github.com/nodejs/node/pull/7575
Reviewed-By: Rod Vagg <rod@vagg.org>
Reviewed-By: Evan Lucas <evanlucas@me.com>
Reviewed-By: Santiago Gimeno <santiago.gimeno@gmail.com>

cjihrig committed:
src: suppress coverity message

Coverity marked a change in 630096b as a constant expression.
However, on platforms where sizeof(int64_t) > sizeof(size_t),
this should not be the case. This commit flags the comparison
as OK to coverity.

PR-URL: https://github.com/nodejs/node/pull/7587
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Michael Dawson <michael_dawson@ca.ibm.com>

Trott committed:
tools: remove unused variable

`oldDirs` is assigned but never used. Remove it.

(This was missed by the linter in previous versions of ESLint but is
flagged by the current version. Updating the linter is contingent on
this change or some similar remedy landing.)

PR-URL: https://github.com/nodejs/node/pull/7594
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>

Trott committed:
test: remove unused var in test-tls-server-verify

`connections` is assigned but never used. Remove it.

(This was missed by the linter in previous versions of ESLint but is
flagged by the current version. Updating the linter is contingent on
this change or some similar remedy landing.)

PR-URL: https://github.com/nodejs/node/pull/7595
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

Trott committed:
test: remove unused var from stream2 test

`writes` is assigned but never used. Remove it.

(This was missed by the linter in previous versions of ESLint but is
flagged by the current version. Updating the linter is contingent on
this change or some similar remedy landing.)

PR-URL: https://github.com/nodejs/node/pull/7596
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>

Trott committed:
test: remove unused var in net-server-try-ports

`connections` is assigned but never used. Remove it.

(This was missed by the linter in previous versions of ESLint but is
flagged by the current version. Updating the linter is contingent on
this change or some similar remedy landing.)

PR-URL: https://github.com/nodejs/node/pull/7597
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>

Trott committed:
test: remove unused vars from http/https tests

Remove handful of variables that are assigned but never used.

(This was missed by the linter in previous versions of ESLint but is
flagged by the current version. Updating the linter is contingent on
this change or some similar remedy landing.)

PR-URL: https://github.com/nodejs/node/pull/7598
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>

Trott committed:
test: remove unused var from child-process-fork

`messageCount` is assigned, but never used. Remove it.

(This was missed by the linter in previous versions of ESLint but is
flagged by the current version. Updating the linter is contingent on
this change or some similar remedy landing.)

PR-URL: https://github.com/nodejs/node/pull/7599
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>

Trott committed:
benchmark: remove unused variables

Remove variables that are assigned but never used.

(This was missed by the linter in previous versions of ESLint but is
flagged by the current version. Updating the linter is contingent on
this change or some similar remedy landing.)

PR-URL: https://github.com/nodejs/node/pull/7600
Reviewed-By: Rod Vagg <rod@vagg.org>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Jeremiah Senkpiel <fishrock123@rocketmail.com>

Trott committed:
tools: update ESLint, fix unused vars bug

Update ESLint to 3.0.0. This includes an enhancement to `no-unused-vars`
such that it finds a few instances in our code base that it did not find
previously (fixed in previous commits readying this for landing).

PR-URL: https://github.com/nodejs/node/pull/7601
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Roman Reiss <me@silverwind.io>

Trott committed:
doc: add benchmark who-to-CC info

Add benchmark to "Who to CC". Also, alphabetized the only
non-alphabetized subsystem.

PR-URL: https://github.com/nodejs/node/pull/7604
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>

Trott committed:
test: fix flaky test-*-connect-address-family

Skip tests if localhost does not resolve to ::1.

Fixes: https://github.com/nodejs/node/issues/7288
PR-URL: https://github.com/nodejs/node/pull/7605
Reviewed-By: Rod Vagg <rod@vagg.org>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>

bnoordhuis committed:
src: disable stdio buffering

Disable stdio buffering, it interacts poorly with printf() calls from
elsewhere in the program (e.g., any logging from V8.)  Unbreaks among
other things the `--trace_debug_json` switch.

Undoes commit 0966ab99 ("src: force line buffering for stderr"), which
in retrospect is not a proper fix.  Turning on line buffering fixed a
flaky test on SmartOS but the test wasn't failing on other platforms,
where stderr wasn't line-buffered either.  Mark the test flaky again,
it failed once in a run of 333 tries on the smartos-64 buildbot.

Disabling buffering should be safe even when mixed with non-blocking
stdio I/O because libuv goes to great lengths to reopen the tty file
descriptors and falls back to blocking I/O when that fails.

PR-URL: https://github.com/nodejs/node/pull/7610
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Rich Trott <rtrott@gmail.com>

addaleax committed:
test,doc: clarify `buf.indexOf(num)` input range

Hopefully clarify the behaviour of `buffer.indexOf()` and
`buffer.includes()` for numbers in that they will be
truncated to uint8s.

Add tests for that behaviour.

Fixes: https://github.com/nodejs/node/issues/7591
PR-URL: https://github.com/nodejs/node/pull/7611
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

Fishrock123 committed:
test: move parallel/test-tty-* to pseudo-tty/

Refs: https://github.com/nodejs/node/pull/7360
PR-URL: https://github.com/nodejs/node/pull/7613
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

Fishrock123 committed:
test: fix old tty tests

test-tty-wrap hasn’t worked since StreamBase was introduced, I think.
test-tty-stdout-end also happens to works with PipeWrap-s.

Refs: https://github.com/nodejs/node/pull/7360
PR-URL: https://github.com/nodejs/node/pull/7613
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

bnoordhuis committed:
deps: update V8 to 5.1.281.75

Pick up the latest branch-head for V8 5.1.  Introduces a semver-minor
overload of v8::Function::New() for use by v8_inspector.

Refs: https://github.com/nodejs/node/pull/7586
PR-URL: https://github.com/nodejs/node/pull/7615
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Fedor Indutny <fedor.indutny@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Myles Borins <myles.borins@gmail.com>

thefourtheye committed:
tools: cleanup no-build and build-only options

As the `no-build` and `build-only` options are not used anymore, they
can be safely removed.

PR-URL: https://github.com/nodejs/node/pull/7620
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>

thefourtheye committed:
tools: fix broken format string

The format specifier is incomplete and without this the program will
fail at runtime, with "incomplete format" error.

PR-URL: https://github.com/nodejs/node/pull/7620
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>

princejwesley committed:
repl: Mitigate vm #548 function redefinition issue

```js
node 🙈 ₹ git:(upstream ⚡ repl-tmp-548) ./node
> function name() { return "node"; };
undefined
> name()
'node'
> function name() { return "nodejs"; };
undefined
> name()
'nodejs'
>
```

Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Lance Ball <lball@redhat.com>

Trott committed:
tools: enforce JS brace style with linting

Enable `brace-style` in ESLint.

Ref: https://github.com/nodejs/node/pull/7094#discussion_r70149215
PR-URL: https://github.com/nodejs/node/pull/7630
Reviewed-By: Brian White <mscdex@mscdex.net>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Roman Reiss <me@silverwind.io>
Reviewed-By: Trevor Norris <trev.norris@gmail.com>
Reviewed-By: Rod Vagg <rod@vagg.org>

Trott committed:
lib,benchmark,test: implement consistent braces

This change is in preparation for lint-enforced brace style.

PR-URL: https://github.com/nodejs/node/pull/7630
Reviewed-By: Brian White <mscdex@mscdex.net>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Roman Reiss <me@silverwind.io>
Reviewed-By: Trevor Norris <trev.norris@gmail.com>
Reviewed-By: Rod Vagg <rod@vagg.org>

ChALkeR committed:
doc: various documentation formatting fixes

* Fix markdown code sample in releases.md, it was <a id="x.y.x></a>"
* Fix some markdown errors, e.g. in changelogs
* Fix broken defs links, e.g. in domain-postmortem.md
* Fix other broken refs, by addaleax
* Add links to some defs that were present but not linked to
* Remove dead defs
* Move defs to the bottom (one file affected)
* Add language indicators to all code blocks, using `txt` when no
specific language could be chosen
* Some minor formatting changes (spaces, ident, headings)

PR-URL: https://github.com/nodejs/node/pull/7637
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Roman Reiss <me@silverwind.io>

Commit:
build: export more openssl symbols on Windows

This exports even more openssl symbols when building
on Windows. SSL_set_fd is one example of added symbol.

PR-URL: https://github.com/nodejs/node/pull/7576
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Roman Reiss <me@silverwind.io>

Trott committed:
tools: increase lint coverage

Extend linting to tools/license2rtf.js and any other JS that gets added
to the `tools` directory by default.

This incidentally simplifies lint invocation.

PR-URL: https://github.com/nodejs/node/pull/7647
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Roman Reiss <me@silverwind.io>
Reviewed-By: Сковорода Никита Андреевич <chalkerx@gmail.com>

thefourtheye committed:
doc: link and highlight Object.assign

PR-URL: https://github.com/nodejs/node/pull/7670
Reviewed-By: Brian White <mscdex@mscdex.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Jeremiah Senkpiel <fishrock123@rocketmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>

evanlucas committed:
doc: fix util.deprecate() example

The arguments object is not created for arrow functions so the example
was incorrect.

PR-URL: https://github.com/nodejs/node/pull/7674
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

cjihrig committed:
test: add common.rootDir

A few of the child process tests can be simplified by computing
the OS specific root directory in common and then accessing that
value.

PR-URL: https://github.com/nodejs/node/pull/7685
Reviewed-By: Roman Reiss <me@silverwind.io>

silverwind committed:
tools: consistent .eslintrc formatting

All quotes in .eslintrc were unnecessary and inconsistently placed
across the file. Additionally, format the globals to be consistent
with the style of whitespace and sorted them alphabetically.

PR-URL: https://github.com/nodejs/node/pull/7691
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>

Trott committed:
test: improve error message in test-tick-processor

Provide additional information about values that indicate test failed.

PR-URL: https://github.com/nodejs/node/pull/7693
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Brian White <mscdex@mscdex.net>

cjihrig committed:
test: cleanup IIFE tests

A number of test files use IIFEs to separate distinct tests from
each other in the same file. The project has been moving toward
using block scopes and let/const in favor of IIFEs. This commit
moves IIFE tests to block scopes. Some additional cleanup such
as use of strictEqual() and common.mustCall() is also included.

PR-URL: https://github.com/nodejs/node/pull/7694
Reviewed-By: Santiago Gimeno <santiago.gimeno@gmail.com>

cjihrig committed:
cluster: remove bind() and self

This commit removes the use of self and bind() from the cluster
module in favor of arrow functions.

PR-URL: https://github.com/nodejs/node/pull/7710
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Santiago Gimeno <santiago.gimeno@gmail.com>
Reviewed-By: Minwoo Jung <jmwsoft@gmail.com>

pmq20 committed:
win,msi: add zh-CN translations for the installer

PR-URL: https://github.com/nodejs/node/pull/2569
Reviewed-By: Jackson Tian <shvyo1987@gmail.com>
Reviewed-By: Wexpo Lyu <laosb@lns.unihubs.com>
Reviewed-By: Yiyu He <heyiyu.deadhorse@gmail.com>
Reviewed-By: Yorkie Liu <yorkiefixer@gmail.com>

cjihrig committed:
test: s/assert.fail/common.fail as appropriate

Many tests use assert.fail(null, null, msg) where it would be
simpler to use common.fail(msg). This is largely because
common.fail() is fairly new. This commit makes the replacement
when applicable.

PR-URL: https://github.com/nodejs/node/pull/7735
Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Jeremiah Senkpiel <fishrock123@rocketmail.com>

Commit:
doc: fix typo in stream doc

Fix small typo in Buffering section of stream doc.
PR-URL: https://github.com/nodejs/node/pull/7738
Reviewed-By: Brian White <mscdex@mscdex.net>
Reviewed-By: Sakthipriyan Vairamani <thechargingvolcano@gmail.com>
Reviewed-By: Michaël Zasso <mic.besace@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>

saadq committed:
doc: fix inconsistencies in code style

Adds missing semicolons, removes extra white space, and properly indents
various code snippets in the documentation.

Reviewed-By: Evan Lucas <evanlucas@me.com>
Reviewed-By: targos - Michaël Zasso <mic.besace@gmail.com>
PR-URL: https://github.com/nodejs/node/pull/7745

kunalspathak committed:
Merge 'nodejs/master' into chakracore-master

Sync nodejs/master at 6510eb5ddc802dcd25713c2b3cbb82711c94d075 as of 2016-07-16

PR-URL: nodejs/node-chakracore#98
Reviewed-By: Jianchun Xu <Jianchun.Xu@microsoft.com>

Commit:
chakrashim: update chakracore (2016-07-22)

Update chakracore to linux HEAD at
https://github.com/Microsoft/ChakraCore/commit/814d084

PR-URL: https://github.com/nodejs/node-chakracore/pull/100
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>

Commit:
Merge remote-tracking branch 'chakracore-master' into xplat

PR-URL: https://github.com/nodejs/node-chakracore/pull/100
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>

Commit:
chakrashim: use updated chakracore

Use new ChakraCoreVersion.h to expose chakracore version.

Use new String APIs for basic string encoding conversions.

Unit test on Ubuntu: `[01:35|% 100|+ 1162|-  31]`

PR-URL: https://github.com/nodejs/node-chakracore/pull/100
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>

kunalspathak commented on deps/chakrashim/core/lib/Jsrt/ChakraCommon.h:
```diff
@@ -2646,6 +2646,49 @@ typedef UINT32 DWORD;
         JsStringFree(
             _In_ char* stringValue);
 
+    // -------- Experimental new String APIs ----------------------------------
```
So are these APIs already in chakracore's master branch?

jianchun replied:
No. I'll push to a temp branch for the time being as these are incubating APIs.

kunalspathak commented on deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:
```diff
@@ -3433,6 +3433,178 @@ CHAKRA_API JsStringFree(_In_ char* stringValue)
+    const SrcChar* end = src + count;
+    while (src < end)
+    {
+        *dst++ = static_cast<DstChar>(*src++);
```
Curious, what happens when `sizeof(SrcChar) > sizeof(DstChar)`? I see below that you are using this method where src is char16 and dst is char. Don't we miss out 8 bits in casting?

jianchun replied:
This is to support v8 String's "OneByte" (latin1) API. Yes, in that case the higher 8 bits are discarded (by design). If the chars are really latin1 this is not a problem. If the chars are not, reading OneByte result is wrong anyway.

kunalspathak replied:
> If the chars are not, reading OneByte result is wrong anyway.

So are you saying `JavascriptString` passed to `JsWriteString` should always be latin1?

jianchun replied:
It is under developers' control, and the behavior would be by design. Yes, supposedly it is only meaningful if the String is latin1 string.

kunalspathak commented on deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:
```diff
@@ -3433,6 +3433,178 @@ CHAKRA_API JsStringFree(_In_ char* stringValue)
+    JsErrorCode errorCode = JsStringToPointer(value, &str, &strLength);
+    if (errorCode != JsNoError)
+    {
+        return errorCode;
```
can you correct the indentation and run lint?

kunalspathak commented on deps/chakrashim/core/lib/Jsrt/Jsrt.cpp:
```diff
@@ -3433,6 +3433,178 @@ CHAKRA_API JsStringFree(_In_ char* stringValue)
+    return JsNoError;
+}
+
+CHAKRA_API JsWriteStringUtf8(
```
Can we rename methods `JsWrite*` to something like `JsWrite*ToBuffer`?

jianchun replied:
These names are unstable yet, could change in future. Feeling adding that is more clear but too long.

kunalspathak commented on deps/chakrashim/chakracore.gyp:
```diff
@@ -114,6 +115,8 @@
         'library_dirs': [ '<(PRODUCT_DIR)' ],
         'conditions': [
           ['OS=="win"', {
+            'libraries': [
```
why is this needed?

jianchun replied:
I forgot...

kunalspathak commented on deps/chakrashim/src/v8v8.cc:
```diff
@@ -18,11 +18,15 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 // IN THE SOFTWARE.
 
+#include <algorithm>
```
nit: don't need this i suppose?

jianchun replied:
It is needed by `std::remove` used in this file. I moved it to front (recall cpplint complains about similar styles).

kunalspathak commented on deps/chakrashim/src/v8string.cc:
```diff
@@ -82,98 +82,57 @@ int String::Utf8Length() const {
-  if (str.From(ref) != JsNoError) {
-    // error
-    return 0;
+int String::Write(uint16_t *buffer, int start, int length, int options) const {
```
does the API says to write into buffer starting from offset `start` and `length` characters?

jianchun replied:
https://github.com/nodejs/node-chakracore/blob/chakracore-master/deps/v8/include/v8.h#L2157

```
   * \param start The starting position within the string at which
   * copying begins.
   * \param length The number of characters to copy from the string.
```

jianchun replied:
Hmm, I am unsure what it does exactly now. Will do some tests to find out later. (Maybe create a bug on this.)

jianchun replied:
Reread old code, should be start in source string. I'll see if I can fix this in time, or create a future bug.

Good catch, thanks!

kunalspathak commented on deps/chakrashim/src/v8string.cc:
```diff
@@ -82,98 +82,57 @@ int String::Utf8Length() const {
-    buffer[count++] = L'\0';
+  size_t count = 0;
+  if (JsWriteString((JsValueRef)this,
+    (char*)buffer + start, length, &count) == JsNoError) {
```
nit: don't need casting  of `(char*)` to buffer + start, right?

kunalspathak replied:
nvm, re-reading `JsWriteString` API. Can you atleast wrap it? `(char*)(buffer + start)` ?

Commit:
chakrashim: add chakracore experimental String APIs

Add chakracore experimental String APIs to support some basic encoding
conversions.

PR-URL: https://github.com/nodejs/node-chakracore/pull/100
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>
